components/director/internal/domain/formationconstraint/service.go

```diff
@@ -99,6 +101,54 @@ func (s *service) ListByFormationTemplateID(ctx context.Context, formationTempla
 	return formationConstraints, nil
 }
 
+// ListByFormationTemplateIDs lists all formation constraints associated with the formation templates
+func (s *service) ListByFormationTemplateIDs(ctx context.Context, formationTemplateIDs []string) ([][]*model.FormationConstraint, error) {
+	constraintRefs, err := s.formationTemplateConstraintReferenceRepo.ListByFormationTemplateIDs(ctx, formationTemplateIDs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while listing Formation Constraint References for FormationTemplates with IDs: %q", formationTemplateIDs)
+	}
+
+	constraintIDs := make([]string, 0, len(constraintRefs))
+	formationTemplateIDToConstraintIDsMap := make(map[string][]string, len(constraintRefs))
+
+	for _, cr := range constraintRefs {
+		constraintIDs = append(constraintIDs, cr.ConstraintID)
+
+		formationTemplateIDToConstraintIDsMap[cr.FormationTemplateID] = append(formationTemplateIDToConstraintIDsMap[cr.FormationTemplateID], cr.ConstraintID)
+	}
+
+	formationConstraints, err := s.repo.ListByIDsAndGlobal(ctx, constraintIDs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while listing Formation Constraints by IDs %q and the global ones", formationTemplateIDs)
+	}
+
+	globalConstraints := make([]*model.FormationConstraint, 0, len(formationConstraints))
+	attachedConstraintsMap := make(map[string]*model.FormationConstraint, len(formationConstraints))
+
+	for _, constraint := range formationConstraints {
+		if constraint.ConstraintScope == model.GlobalFormationConstraintScope {
+			globalConstraints = append(globalConstraints, constraint)
+		} else {
+			attachedConstraintsMap[constraint.ID] = constraint
+		}
+	}
+
+	formationConstraintsPerFormationTemplate := make([][]*model.FormationConstraint, len(formationTemplateIDs))
```

Comment:
To me it seems the logic in this function(`ListByFormationTemplateIDs`) is unnecessarily complex. 

E.g. here you break up into smaller units the `formationConstraints` -  into "global" and "non-global"(which are stored into a new map[constrainID][constraintValue]. And in the for cycle below you combine them again. What's the idea behind this? In the end we combine them again?

Moreover, the final structure is `[][]*model.FormationConstraint` but in the resolver we're interested only of the second part(`[]*model.FormationConstraint`) and the indexes are skipped and not used - `for _, formationConstraints := range formationConstraintsPerFormation {}`? Why we need [][].. when we don't use it

Could we simplify the whole logic to:
```
func (s *service) ListByFormationTemplateIDs(ctx context.Context, formationTemplateIDs []string) ([]*model.FormationConstraint, error) {
	constraintRefs, err := s.formationTemplateConstraintReferenceRepo.ListByFormationTemplateIDs(ctx, formationTemplateIDs)
	if err != nil {
		return nil, errors.Wrapf(err, "while listing Formation Constraint References for FormationTemplates with IDs: %q", formationTemplateIDs)
	}

	constraintIDs := make([]string, 0, len(constraintRefs))
	for _, cr := range constraintRefs {
		constraintIDs = append(constraintIDs, cr.ConstraintID)
	}

	formationConstraints, err := s.repo.ListByIDsAndGlobal(ctx, constraintIDs)
	if err != nil {
		return nil, errors.Wrapf(err, "while listing Formation Constraints by IDs %q and the global ones", formationTemplateIDs)
	}

	return formationConstraints, nil
}
```

Reply:
We discussed it offline. To sum it up: the structure is an array of arrays because of the dataloader(that's how it works). 
And in the result, at the index 0 are the formation constraints for the formation template ID at index 0 from the input formationTemplateIDs; at index 1 are the FCs for the FT at index 1, etc. That's how the dataloaders aggregate which constraints are for which formation template.